src/napari_spatialdata/_viewer.py

```diff
@@ -0,0 +1,164 @@
+from __future__ import annotations
+
+from typing import TYPE_CHECKING
+
+import numpy as np
+from anndata import AnnData
+from loguru import logger
+from multiscale_spatial_image import MultiscaleSpatialImage
+from napari import Viewer
+from napari.layers import Labels, Points, Shapes
+from napari.utils.notifications import show_info
+
+from napari_spatialdata.utils._utils import _get_transform, _swap_coordinates
+
+if TYPE_CHECKING:
+    from napari.utils.events.event import Event
+    from spatialdata import SpatialData
+
+
+class SpatialDataViewer:
+    def __init__(self, viewer: Viewer) -> None:
+        self.viewer = viewer
+        self.viewer.layers.events.inserted.connect(self._inherit_metadata)
```

Comment:
Thanks for the updates, @melonora ! I am still a bit concerned about this event connection. If I understand the code correctly, it the attached callback will try to add spatialdata metadata to _any_ layer added to the viewer. Is this correct?

While I can see how that would be desirable for the case where spatialdata is the only library/plugin being used with napari, I would hope that we can design our plugins such that they can be used in conjuction with other plugins and workflows. If we are automatically adding spatialdata metadata to layers added by other plugins or by the user with other intentions, that seems like it could potentially cause issues (e.g., adding spatialdata metadata to a layer where it doesn't make sense to do so).

Instead of connecting the `_inherit_metadata()` method to this event, can we call it in the `add_sdata_*()` methods? Perhaps we can also have a public version that takes a source layer and "clones" the metadata onto a target layer for the case when people want to manually add metadata? What do you think?

Reply:
> Thanks for the updates, @melonora ! I am still a bit concerned about this event connection. If I understand the code correctly, it the attached callback will try to add spatialdata metadata to _any_ layer added to the viewer. Is this correct?

Yes this is correct if there is an active layer. When napari starts and no spatialdata layers have been opened yet, no spatialdata metadata can be added.

> While I can see how that would be desirable for the case where spatialdata is the only library/plugin being used with napari, I would hope that we can design our plugins such that they can be used in conjuction with other plugins and workflows. If we are automatically adding spatialdata metadata to layers added by other plugins or by the user with other intentions, that seems like it could potentially cause issues (e.g., adding spatialdata metadata to a layer where it doesn't make sense to do so).

What kind of case do you have in mind where it makes sense for the user to have a spatialdata object open without wanting an association with that object when opening a new layer? Just to have a clear idea.
> 
> Instead of connecting the `_inherit_metadata()` method to this event, can we call it in the `add_sdata_*()` methods? Perhaps we can also have a public version that takes a source layer and "clones" the metadata onto a target layer for the case when people want to manually add metadata? What do you think?

I think it would be possible by means of having source layer and target layer selected and then using a shortcut. Adding it to the `add_sdata_*()` methods would be more complicated I think. It is specifically the usecase when new layers are added to do annotation on a spatialdata object. How would you distinguish from the case you described above when a user presses the button to add a new layer which should not be associated with the spatialdata object? I think our best option would be to go for a public method triggered by a keyboard shortcut as I described if we decide to allow for layers not to be associated with a spatialdata object.

Another question is also whether we should allow for 2 spatialdata objects to be associated with one layer.

Reply:
> What kind of case do you have in mind where it makes sense for the user to have a spatialdata object open without wanting an association with that object when opening a new layer? Just to have a clear idea.

One example is that some plugins create temporary layers for annotations or displaying visuals. I don't think these should be associated with the spatialdata object. In general, I don't think we should start by automatically adding metadata to all layers. We can add it in the future as necessary, but I don't think we should be careful about adding "magic" at this early stage.

> I think our best option would be to go for a public method triggered by a keyboard shortcut as I described if we decide to allow for layers not to be associated with a spatialdata object.

I agree this is the best approach. We can have a public helper function that adds a layer for annotation (i.e., adds the layer and assigns the spatialdata metadata) and potentially some UI buttons in the future.